### PR TITLE
[IP addresses] Replace the old implementation by the new one

### PR DESCRIPTION
**Pull request by gx740**

### CODE REVIEW

**Comment by seladb** [comment]
I'm going on a vacation and I won't take my laptop with me. I'll review the PR as soon as I get back in the second week of Jan

File: Examples/PcapSplitter/Splitters.h [hunk 1]
```diff
@@ -1,6 +1,10 @@
 #pragma once
 
 #include <LRUList.h>
+
+#if !defined(WIN32) && !defined(WINx64) && !defined(PCAPPP_MINGW_ENV)
+#include <netinet/in.h>
+#endif
```
Comment (seladb) [inline Examples/PcapSplitter/Splitters.h]: why are those lines needed here?

File: Packet++/header/DnsResourceData.h [hunk 2]
```diff
@@ -5,6 +5,7 @@
 #include "IpAddress.h"
 #include <string>
 #include <stdint.h>
+#include <memory>
```
Comment (seladb) [inline Packet++/header/DnsResourceData.h]: why is this include needed?

File: Packet++/src/TcpReassembly.cpp [hunk 3]
```diff
@@ -164,25 +164,18 @@ void TcpReassembly::reassemblePacket(Packet& tcpData)
...
 	// calculate packet's source and dest IP address
-	IPAddress* srcIP = NULL;
-	IPAddress* dstIP = NULL;
-	IPv4Address srcIP4Addr = IPv4Address::Zero;
-	IPv6Address srcIP6Addr = IPv6Address::Zero;
-	IPv4Address dstIP4Addr = IPv4Address::Zero;
-	IPv6Address dstIP6Addr = IPv6Address::Zero;
+	IPAddress srcAddr, dstAddr;
+	IPAddress* srcIP = &srcAddr;
+	IPAddress* dstIP = &dstAddr;
```
Comment (seladb) [inline Packet++/src/TcpReassembly.cpp]: I'm not sure why you still need these 2 pointers here

File: Pcap++/src/PcapLiveDeviceList.cpp [hunk 4]
```diff
@@ -231,7 +229,7 @@ PcapLiveDevice* PcapLiveDeviceList::getPcapLiveDeviceByIp(IPv4Address ipAddr) co
 				continue;
 			}
 
-			if (currAddr->s_addr == ipAddr.toInAddr()->s_addr)
+			if (memcmp(currAddr, ipAddr.toBytes(), sizeof(*currAddr)) == 0)
```
Comment (seladb) [inline Pcap++/src/PcapLiveDeviceList.cpp]: why using `memcmp` here instead of comparing `currAddr->s_addr` with `ipAddr.toUInt()`?

File: Pcap++/src/PcapRemoteDeviceList.cpp [hunk 5]
```diff
@@ -121,7 +119,7 @@ PcapRemoteDevice* PcapRemoteDeviceList::getRemoteDeviceByIP(IPv4Address ip4Addr)
 				continue;
 			}
 
-			if (currAddr->s_addr == ip4Addr.toInAddr()->s_addr)
+			if (memcmp(currAddr, ip4Addr.toBytes(), sizeof(currAddr)) == 0)
```
Comment (seladb) [inline Pcap++/src/PcapRemoteDeviceList.cpp]: why using `memcmp` here instead of comparing `currAddr->s_addr` with `ip4Addr.toUInt()`?

File: Examples/PcapSplitter/Splitters.h (hunk 1 above)
Comment (gx740) [reply Examples/PcapSplitter/Splitters.h]: The following compilation errors occur under Linux:
```
==> Building target: PcapSplitter
Building file: main.cpp
In file included from Splitters.h:8,
                 from SimpleSplitters.h:3,
                 from main.cpp:56:
/home/ssuslov/src/PcapPlusPlus/Dist/header/IpUtils.h:63:2: error: ‘in_addr’ does not name a type; did you mean ‘s6_addr’?
  in_addr* sockaddr2in_addr(struct sockaddr *sa);
  ^~~~~~~
  s6_addr
/home/ssuslov/src/PcapPlusPlus/Dist/header/IpUtils.h:70:2: error: ‘in6_addr’ does not name a type; did you mean ‘s6_addr’?
  in6_addr* sockaddr2in6_addr(struct sockaddr *sa);
  ^~~~~~~~
  s6_addr
/home/ssuslov/src/PcapPlusPlus/Dist/header/IpUtils.h:84:23: error: ‘in_addr’ was not declared in this scope
  uint32_t in_addr2int(in_addr inAddr);
                       ^~~~~~~
/home/ssuslov/src/PcapPlusPlus/Dist/header/IpUtils.h:84:23: note: suggested alternative: ‘s6_addr’
  uint32_t in_addr2int(in_addr inAddr);
                       ^~~~~~~
                       s6_addr
make[1]: *** [Makefile:16: Obj/main.o] Error 1
```

Comment (gx740) [reply Examples/PcapSplitter/Splitters.h]: I could not fix this problem in different way

File: Packet++/header/DnsResourceData.h (hunk 2 above)
Comment (gx740) [reply Packet++/header/DnsResourceData.h]: It's necessary for the following code:

```c++
#if __cplusplus > 199711L || _MSC_VER >= 1800 //Maybe this can be 1600 for VS2010
#define PCPP_SMART_PTR(T) std::unique_ptr<T>
#else
#define PCPP_SMART_PTR(T) std::auto_ptr<T>
#endif
```

File: Packet++/src/TcpReassembly.cpp (hunk 3 above)
Comment (gx740) [reply Packet++/src/TcpReassembly.cpp]: It's a temporary code to make less changes. In the next PR all of these will changed.

File: Pcap++/src/PcapLiveDeviceList.cpp (hunk 4 above)
Comment (gx740) [reply Pcap++/src/PcapLiveDeviceList.cpp]: For IPv6 address the following check is used:
```c++
if (memcmp(currAddr, addrAsArr, sizeof(struct in6_addr)) == 0)
```

I just made the check for IPv4 in the same way. The possible replacement is:
```c++
if (currAddr->s_addr == ipAddr.toUInt())
```

File: Pcap++/src/PcapRemoteDeviceList.cpp (hunk 5 above)
Comment (gx740) [reply Pcap++/src/PcapRemoteDeviceList.cpp]: The same as for `PcapLiveDeviceList.cpp`

File: Pcap++/src/PcapLiveDeviceList.cpp (hunk 4 above)
Comment (gx740) [reply Pcap++/src/PcapLiveDeviceList.cpp]: Should I remove `memcmp`?

Comment (gx740) [reply Pcap++/src/PcapLiveDeviceList.cpp]: `memcmp` was removed

File: Pcap++/src/PcapRemoteDeviceList.cpp (hunk 5 above)
Comment (gx740) [reply Pcap++/src/PcapRemoteDeviceList.cpp]: `memcmp` was removed

**Comment by gx740** [comment]
Can you push the latest changes in `dev` into `ipaddr_refactor`?

File: Packet++/src/TcpReassembly.cpp (hunk 3 above)
Comment (seladb) [reply Packet++/src/TcpReassembly.cpp]: ok sure

File: Examples/PcapSplitter/Splitters.h (hunk 1 above)
Comment (seladb) [reply Examples/PcapSplitter/Splitters.h]: that is weird because you haven't made any other changes in this file...

**Comment by seladb** [comment]
Done. 

Please sync this PR to `ippaddr_refactor` and make sure there are no conflict or errors and then I'll merge the PR

File: Examples/PcapSplitter/Splitters.h (hunk 1 above)
Comment (gx740) [reply Examples/PcapSplitter/Splitters.h]: There is the following code in  `IpUtils.h`:
```c++
#ifdef LINUX
#include <in.h>
#include <arpa/inet.h>
#endif
#ifdef MAC_OS_X
#include <netinet/in.h>
#include <arpa/inet.h>
#endif
#if defined(WIN32) || defined(WINx64) || defined(PCAPPP_MINGW_ENV)
#include <winsock2.h>
#include <ws2tcpip.h>
#endif
#ifdef FREEBSD
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#endif
```

And I think the pre-processor constant `LINUX` is not defined in example for some reason and necessary header files are not included.

**Comment by gx740** [comment]
Done

File: Examples/PcapSplitter/Splitters.h (hunk 1 above)
Comment (seladb) [reply Examples/PcapSplitter/Splitters.h]: I think I managed to fix that!

You can remove `IpUtils.h` from `Splitters.h` and add it to `IPPortSplitters.h` below the `#if` clause, like that:

```cpp
#include "Splitters.h"
#if !defined(WIN32) && !defined(WINx64) //for using ntohl, ntohs, etc.
#include <in.h>
#endif
#include "IpUtils.h"
```

I haven't checked that on Windows or Mac though, just on Linux and it's building fine

Comment (gx740) [reply Examples/PcapSplitter/Splitters.h]: The changes have been made

Comment (seladb) [reply Examples/PcapSplitter/Splitters.h]: Great thanks!

**Comment by seladb** [comment]
Are there any more changes you're planning in this PR or can we merge it?

**Comment by gx740** [comment]
No more changes are planning
